### PR TITLE
Feature/mx 1473 integrate ldap into seq repo

### PR DESCRIPTION
**Pull request by mr-kamran-ali**
# PR Context
- Goal was to integrate LDAP and Organigram in the transformation process 
- If resolved persons has a department, then integrate both ldap resolved person and units in the transformation

# Added
- LDAP and Organigram integration according to mapping

# Changes
- During transformation process check with ldap and organigram to resolve units or persons

### CODE REVIEW

File: mex/seq_repo/transform.py [hunk 1]
```diff
@@ -1,43 +1,77 @@
...
+        unit_stable_target_ids_by_synonym: Unit stable target ids by synonym
+        project_coordinators_merged_ids_by_query_string: Seq Repo Sources project
+                                                        coordinators merged ids
         extracted_primary_source: Extracted primary source
 
     Returns:
         Generator for ExtractedActivity
     """
     theme = seq_repo_activity["theme"][0]["mappingRules"][0]["setValues"]
-    for source in seq_repo_sources.values():
```
Comment (erichesse) [inline mex/seq_repo/transform.py:30]: That was better though

File: mex/seq_repo/transform.py [hunk 2]
```diff
@@ -191,15 +259,21 @@ def transform_seq_repo_access_platform_to_extracted_access_platform(
     ][0]["setValues"]
     title = seq_repo_access_platform["title"][0]["mappingRules"][0]["setValues"]
 
+    contacts = seq_repo_access_platform["contact"][0]["mappingRules"][0]["forValues"]
+
+    resolved_organigram = []
+    for contact in contacts:
+        resolved_organigram.append(unit_stable_target_ids_by_synonym.get(contact))
```
Comment (erichesse) [inline mex/seq_repo/transform.py:266]:
```suggestion
    resolved_organigram = [unit_stable_target_ids_by_synonym.get(contact) for contact in contacts]
```

**Review by erichesse — COMMENTED** [review]
Looks very good. Just some smaller issues and refactoring of some duplicate code left

File: mex/seq_repo/extract.py [hunk 3]
```diff
@@ -22,3 +25,31 @@ def extract_sources() -> Generator[SeqRepoSource, None, None]:
         data = json.load(file)
         for item in data:
             yield SeqRepoSource.model_validate(item)
+
+
+@watch
+def extract_source_project_coordinator(
+    seq_repo_sources: dict[str, SeqRepoSource],
+) -> Generator[LDAPPersonWithQuery, None, None]:
```
[the remark this answers is not in the document]
Comment (mr-kamran-ali) [reply mex/seq_repo/extract.py:33]: Few minor issues with this approach: 

- I have to declare empty list and keep adding items in it in the loop and return the whole list at the end. 
- @watch does not work with list
- it breaks consistency with ldap extraction of other extractors

File: mex/seq_repo/transform.py [hunk 4]
```diff
@@ -125,16 +169,40 @@ def transform_seq_repo_resource_to_extracted_resource(
...
+                    and query_ldap.person.departmentNumber
+                ):
+                    if query_ldap.person.sAMAccountName.lower() == pc.lower():
+                        unit = unit_stable_target_ids_by_synonym.get(
+                            query_ldap.person.departmentNumber
+                        )
+                        if unit and unit not in units_in_charge:
+                            units_in_charge.append(unit)
+        if not units_in_charge or not project_coordinators_ids:
+            continue
```
[the remark this answers is not in the document]
Comment (mr-kamran-ali) [reply mex/seq_repo/transform.py]: this was refactored in 1ea867d253d8013141bdb063b69a9628b3be20d3

File: tests/seq_repo/conftest.py [hunk 5]
```diff
@@ -371,3 +424,77 @@ def extracted_mex_distribution_dict(
...
         for distribution in extracted_mex_distributions
     }
+
+
+@pytest.fixture
+def seq_repo_source_project_coordinators(
+    seq_repo_latest_sources: dict[str, SeqRepoSource],
+) -> list[LDAPPersonWithQuery]:
+    """Extract source project coordinators."""
+    return list(extract_source_project_coordinator(seq_repo_latest_sources))
```
[the remark this answers is not in the document]
Comment (mr-kamran-ali) [reply tests/seq_repo/conftest.py]: data isolated in c0d3c4f940730d1187ffaf1528ce346378027879

File: tests/seq_repo/conftest.py [hunk 6]
```diff
@@ -371,3 +424,77 @@ def extracted_mex_distribution_dict(
...
+@pytest.fixture
+def unit_stable_target_ids_by_synonym(
+    extracted_primary_sources: dict[str, ExtractedPrimarySource],
+) -> dict[str, MergedOrganizationalUnitIdentifier]:
+    """Extract the dummy units and return them grouped by synonyms."""
+    organigram_units = extract_organigram_units()
+    mex_organizational_units = transform_organigram_units_to_organizational_units(
+        organigram_units, extracted_primary_sources["organigram"]
+    )
+    return get_unit_merged_ids_by_synonyms(mex_organizational_units)
```
[the remark this answers is not in the document]
Comment (mr-kamran-ali) [reply tests/seq_repo/conftest.py]: data hardcoded in 3b3e6898fe84dc0ebc66ff90b8064501fa85bb1a

File: mex/seq_repo/transform.py [hunk 7]
```diff
@@ -1,43 +1,77 @@
...
+
+            for query in seq_repo_source_project_coordinators:
+                query_ldap: LDAPPersonWithQuery = query
+                if (
+                    query_ldap.person.sAMAccountName
+                    and query_ldap.person.departmentNumber
+                ):
+                    if query_ldap.person.sAMAccountName.lower() == pc.lower():
+                        unit = unit_stable_target_ids_by_synonym.get(
+                            query_ldap.person.departmentNumber
```
[the remark this answers is not in the document]
Comment (erichesse) [reply mex/seq_repo/transform.py]: sorry, If my suggestion was misleading, I meant to declare the variables in the if clause, and then reuse them in the next lines as they actually are used twice

**Comment by erichesse** [comment]
Looks good, just one misunderstanding to resolve.

**Review by erichesse — COMMENTED** [review]
Almost done!

**Comment by erichesse** [comment]
> Almost done!

Ah, you should also update the changelog

File: mex/seq_repo/transform.py [hunk 8]
```diff
@@ -191,15 +238,61 @@ def transform_seq_repo_access_platform_to_extracted_access_platform(
...
+    for pc in project_coordinators:
+        person_merged_id = project_coordinators_merged_ids_by_query_string.get(pc)
+        if person_merged_id:
+            project_coordinators_ids.append(person_merged_id[0])
+
+        for query in seq_repo_source_resolved_project_coordinators:
+            query_ldap: LDAPPersonWithQuery = query
+            sam_account_name = query_ldap.person.sAMAccountName
+            department_number = query_ldap.person.departmentNumber
+            if sam_account_name and department_number:
```
Comment (erichesse) [inline mex/seq_repo/transform.py:293]:
```suggestion
            if (sam_account_name := query_ldap.person.sAMAccountName) and (
                department_number := query_ldap.person.departmentNumber
            ):
```
Almost :)

**Review by erichesse — COMMENTED** [review]
Close, just declare it in the clause directly

**Review by erichesse — APPROVED** [review]
Great! ready to ship!

**Pull request merged**